copy/copy.go

```diff
@@ -816,7 +926,32 @@ func (c *copier) copyBlobFromStream(ctx context.Context, srcStream io.Reader, sr
 	// Note that for this check we don't use the stronger "validationSucceeded" indicator, because
 	// dest.PutBlob may detect that the layer already exists, in which case we don't
 	// read stream to the end, and validation does not happen.
-	digestingReader, err := newDigestingReader(srcStream, srcInfo.Digest)
+
+	var decrypted bool
+	var err error
+	if manifest.IsEncryptedLayer(srcInfo) {
+		if c.decryptConfig == nil {
+			return types.BlobInfo{}, ErrDecryptParamsMissing
+		}
+
+		newDesc := ocispec.Descriptor{
+			Annotations: srcInfo.Annotations,
+		}
+
+		var d digest.Digest
+		srcStream, d, err = ocicrypt.DecryptLayer(c.decryptConfig, srcStream, newDesc, false)
```

Comment:
(Why is it OK for the generic `copy` code to hard-code OCI-specific format implementations?

It _might_ be, but it’s not trivially clear.

If the code hard-codes OCI-specific handling here, the user-visible parameters in `SystemContext` should probably also explicitly say OCI in the name. … but then, can we legitimately name that OCI without being OCI-accepted?)

Reply:
> I think having an encryption abstraction is fine, but at the moment, there isn't an alternative implementation, so defining the interface for the abstraction will most likely end up in something similar or identical to what we already have.

Not if the encryption systems are different, and require different formats/kinds of keys and other parameters. We would probably end up with a `SystemContext`-like way to hide the differences… or maybe require the caller to pass an encryption-system-specific key set object that has already been initialized, and somehow match it to the mechanisms supported by various manifest types.

The “generic” interface like that, however it ends up looking, would be quite different from the current one. Hence the suggestion that the API parameters to this encryption mechanism should explicitly say something like “OCI”, to be differentiated from the hypothetical future one.

Reply:
Alright sounds good! - so for now I will add OCI prefixed naming for encryption specific variables.